changefeedccl: add network tracking

andrewbaptist opened the pull request:
Adds network tracing infrasturcture to changefeeds. This commit adds the
metrics but does not populate them for any of the existing changefeeds.

Part of: https://github.com/cockroachdb/cockroach/issues/130097

Epic: none

Release note (ops change): This commit adds two metrics
changefeed.network.bytes_in and changefeed.network.bytes_out. These
metrics track the number of bytes sent by the individual changefeeds to
different sinks.

Commit:
changefeedccl: add network tracking

Adds network tracing infrasturcture to changefeeds. This commit adds the
metrics but does not populate them for any of the existing changefeeds.

Part of: #130097

Epic: none

Release note (ops change): This commit adds two metrics
changefeed.network.bytes_in and changefeed.network.bytes_out. These
metrics track the number of bytes sent by the individual changefeeds to
different sinks.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129151)
<!-- Reviewable:end -->

andrewbaptist commented:
This is still a WIP - but I think I have a working model for changing all the consumers/producers... testing this is going to be a little tricky though :)

asg0451 commented on pkg/ccl/changefeedccl/sink_kafka_v2.go:
```diff
@@ -340,6 +342,7 @@ func makeKafkaSinkV2(
 	settings *cluster.Settings,
 	mb metricsRecorderBuilder,
 	knobs kafkaSinkV2Knobs,
+	netMetrics *cidr.NetMetrics,
```
instead of passing it in separately, pls put this in the metricsRecorder/Builder interface, which presumably is why you put it in sliMetrics in the first place

andrewbaptist replied:
Fixed - I had meant to use this intead.

asg0451 commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -88,6 +89,10 @@ type AggMetrics struct {
+
+	// TODO(baptist): This doesn't really belong here, but is easier than
+	// threading the NetMetrics through all the other places.
+	NetMetrics *cidr.NetMetrics
```
agree with this comment. can you make a gh issue to clean this up and reference it in this comment pls

andrewbaptist replied:
Filed #130358 to track this.

asg0451 commented on pkg/ccl/changefeedccl/sink_kafka_v2.go:
```diff
@@ -419,14 +424,18 @@ func buildKgoConfig(
 		}
-		opts = append(opts, kgo.DialTLSConfig(tlsCfg))
+
+		dialer := &net.Dialer{Timeout: 10 * time.Second}
```
this looks ok but pls add tests for both of these cases

andrewbaptist replied:
I might need some help with this. I think this code is only tested if we are connecting to a "real" endpoint.  I wasn't clear if the unit tests did this, although I do see `kafka://localhost:9092` in the tests, so they might... If we can't unit test this, is there a roachtest that sets this up that we could check the stat on?

asg0451 replied:
yeah the unit tests don't connect to real kafka. the roachtests do. but we also have a test in sink_kafka_v2_test.go that tests options being set. ideally you'd add it to both of these but given the time constraints maybe just the latter.

andrewbaptist replied:
I pulled the kafka changes to a separate PR, so this comment is stale, but I will look at ways to test on that PR.

andyyang890 commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -736,9 +755,21 @@ var (
 	}
+	metaNetworkBytesIn = metric.Metadata{
+		Name:        "changefeed.network.bytes_in",
+		Help:        "The number of received from the network",
```
nit: s/received/bytes received/ and also add "by changefeeds" to the end

andyyang890 commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -736,9 +755,21 @@ var (
+	}
+	metaNetworkBytesOut = metric.Metadata{
+		Name:        "changefeed.network.bytes_out",
+		Help:        "The number of bytes sent over the network",
```
nit: add "by changefeeds" to the end

andrewbaptist replied:
Done

Commit:
util: add additional wrapping methods to cidr

For changefeeds we need some additional network wrapping methods. This
commit also adds testing to Wrap, WrapTLS and WrapDialer.

Part of: #130097

Epic: none

Release note: None

andyyang890 reviewed (approved):
Please address Miles' nits on the other PR in this PR:
- https://github.com/cockroachdb/cockroach/pull/130437#discussion_r1752658439
- https://github.com/cockroachdb/cockroach/pull/130437#discussion_r1752658769

asg0451 commented on pkg/ccl/changefeedccl/metrics.go:
```diff
@@ -88,6 +89,10 @@ type AggMetrics struct {
 		sliMetrics map[string]*sliMetrics
 	}
+
+	// TODO(baptist): This doesn't really belong here, but is easier than
```
see my comment in the other pr and above `TODO(#issue number)`

asg0451 commented on pkg/util/cidr/cidr.go:
```diff
@@ -411,6 +428,47 @@ func (m *NetMetrics) WrapTLS(dial DialContext, tlsCfg *tls.Config, labels ...str
+
+// WrapDialer returns a Dialer that wraps the connection with metrics. If the
+// underlying library exposes an ability to replace the DialContext, you should
+// use WrapDial instead of this function.
```
```suggestion
// use Wrap instead of this function.
```

asg0451 reviewed (approved):
lgtm % nits

andrewbaptist commented:
TFTR!

bors r=asg0451

andrewbaptist commented:
The full set of changes that need to be backported for CDC are:
#130424, #130428, #130434, #130516, #130437
